src/Data/Vec/Properties.agda

```diff
@@ -691,8 +696,209 @@ foldr-fusion : ∀ {A : Set a}
 foldr-fusion {B = B} {f} e {C} h fuse =
   foldr-universal C _ _ refl (λ x xs → fuse x (foldr B f e xs))
 
-idIsFold : ∀ {n} → id ≗ foldr (Vec A) {n} _∷_ []
-idIsFold = foldr-universal _ _ id refl (λ _ _ → refl)
+idIsFoldr : ∀ {n} → id ≗ foldr (Vec A) {n} _∷_ []
+idIsFoldr = foldr-universal _ _ id refl (λ _ _ → refl)
+
+foldr-[] : ∀ {A : Set a} (B : ℕ → Set b)
+           (f : ∀ {n} → A → B n → B (suc n)) {e} → foldr B f e [] ≡ e
+foldr-[] B f = refl
+
+foldr-++ : ∀ {A : Set a} (B : ℕ → Set b)
+           (f : ∀ {n} → A → B n → B (suc n)) {e} →
+           ∀ {m n} {xs} {ys} →
+           foldr B {m + n} f e (xs ++ ys) ≡ foldr (B ∘ (_+ n)) f (foldr B {n} f e ys) xs
+foldr-++ B f {xs = []}     = refl
+foldr-++ B f {xs = x ∷ xs} = P.cong (f x) (foldr-++ B f {xs = xs})
+
+foldl-universal : ∀ {A : Set a} (B : ℕ → Set b)
+                  (f : ∀ {n} → B n → A → B (suc n)) {e}
+                  (h : ∀ {c} (C : ℕ → Set c) →
+                       (g : ∀ {n} → C n → A → C (suc n)) → C zero →
+                       ∀ {n} → Vec A n → C n) →
+                  (∀ {c} {C} {g : ∀ {n} → C n → A → C (suc n)} e →
+                   h {c} C g e [] ≡ e) →
+                  (∀ {c} {C} {g : ∀ {n} → C n → A → C (suc n)} e →
+                   ∀ {n} x →
+                   (h {c} C g e {suc n}) ∘ (x ∷_) ≗ h (C ∘ suc) (λ {n} → g {suc n}) (g e x)) →
+                  ∀ {n} → h B f e ≗ foldl B {n} f e
+foldl-universal B f {e} h base step []       = base e
+foldl-universal B f {e} h base step (x ∷ xs) = begin
+  h B f e (x ∷ xs)
+    ≡⟨ step e x xs ⟩
+  h (B ∘ suc) (λ {n} → f {suc n}) (f e x) xs
+    ≡⟨ foldl-universal (B ∘ suc) (λ {n} → f {suc n}) {f e x} h base step xs ⟩
+  foldl (B ∘ suc) (λ {n} → f {suc n}) (f e x) xs
+    ≡⟨⟩
+  foldl B f e (x ∷ xs)
+    ∎
+  where open P.≡-Reasoning
+
+foldl-[] : ∀ {A : Set a} (B : ℕ → Set b)
+           (f : ∀ {n} → B n → A → B (suc n)) {e} → foldl B f e [] ≡ e
+foldl-[] B f = refl
+
+foldl-fusion : ∀ {A : Set a}
+               {B : ℕ → Set b} {C : ℕ → Set c}
```

Comment:
I suspect `B` and `C` need to explicit, unless you have evidence to the contrary?

Reply:
It turns out that all uses of `foldl-fusion` seem able to infer both `B` and `C`: in the only non-trivial instance, in `Properties.Heterogeneous`, they get inferred by supplying an explicit type signature for the fused function `h` (and this might well be a useful pattern to bear in mind when invoking `foldl-fusion h`...?)

Reply:
And indeed, arguments `d` and `e` to this function can also be made implicit.

Reply:
Thanks for investigating!